### ISSUE TITLE
[Feature] Predict Average of Trees

### ISSUE THREAD

**Issue by Laurae2** [Microsoft/LightGBM#384]
Is there a way to predict the average of boosted trees of LightGBM?

I am trying to implement a stochastic optimizer based on LightGBM, by doing the following:

1. Creating a 2000-tree boosting model with `learning_rate = 1e-5` (to make it "not learn", but with enough precision to reverse engineer each tree), taking each 10% of samples only
2. Averaging each tree predictions to reduce noise, using hacky predictions in R (reverse engineering original tree predictions) <= **this is what I would like to make simple**
3. Averaging each leaf prediction to reduce noise <= this is already provided
4. Go back to step 1 using the predictions from step 2 and 3 (similar to Cascade Forest)

I can't use boosting because it overfits severely, like my [hybrid cross-entropy optimizer](https://github.com/Laurae2/Laurae/blob/master/R/ExtraOpt.R) using boosting.

I noticed we can not use a learning rate of 0.0 in LightGBM. Would there be a way to force using average, if the `learning_rate` is zero? (or use a prediction parameter to get the average?).

Also, I think this might fix issue #47 at the same time, in a much simpler way (no modification of any model - only of the predictor).

**Comment by guolinke** [comment]
@Laurae2 
How to ensure the 2000 trees in step.1 not same ? Some random settings?

**Comment by guolinke** [comment]
I think it is like the random forest. 

@wxchan Did you have any updates on RF ?

We can set ```num_iterations``` and ```num_tree_per_iteration``` at the same time.

**Comment by Laurae2** [comment]
@guolinke Yes, I use bagging 10% of samples every iteration.

```r
model <- lgb.train(dparam,
                   dtrain,
                   2000,
                   dval,
                   min_data = 1,
                   learning_rate = 1e-5,
                   early_stopping_rounds = 2000,
                   bagging_fraction = 0.1,
                   bagging_freq = 1,
                   bagging_seed = 1)
```

It find it is very strong (for leaf prediction) when used locally.

```r
  0   1   2   3   4   5   6   7   8 
322 165 159 161 160 161 162 164 157 
```

![image](https://cloud.githubusercontent.com/assets/9083669/24752858/0e83b896-1ad0-11e7-96a0-6abcaa3cdc85.png)

![image](https://cloud.githubusercontent.com/assets/9083669/24752868/12b29752-1ad0-11e7-8727-7ab1995788e5.png)

![image](https://cloud.githubusercontent.com/assets/9083669/24752874/17a08e22-1ad0-11e7-8933-d5e0da1f5055.png)


I can add this example to R examples if you want.

**Comment by Laurae2** [comment]
> We can set num_iterations and num_tree_per_iteration at the same time.

I think `num_tree_per_iteration` is not yet exposed for training. If I set to a different value than 1, I get back 1:

```
tree
num_class=1
num_tree_per_iteration=1
label_index=0
max_feature_idx=125
objective=regression
feature_names=cap-shape=bell
```

**Comment by Laurae2** [comment]
Duplicate of issue #47, closing
